### DISCUSSION TITLE
Deep assignment statement

### DISCUSSION THREAD

**Discussion by fmontesi** [jolie/jolie#419] [Ideas (Features, Fixes, etc.)]
Following up on the discussion in #273, here is a proposal of what we could do (partially discussed with @klag already):

- Make `<<` deprecated in tree literals (as in, a warning will be issued by the interpreter).
- Add a new "deep assignment" statement `path := expr` that behaves as expected. For example, `x := expr` replaces whatever is in `x` with the entire tree that results from evaluating `expr`. Differently from `<<`, `x := y` is always interpreted as `x[0] := y[0]`: `:=` does not copy vectors. Copying vectors will have to be done with either for loops or simply by invoking an identity operation, e.g., `id@vectors`.
- Add an operation to the `Values` library called `merge`, which does the job of merging two trees together.

An alternative to introducing `:=` would be to just change the semantics of `<<` to a deep assignment. This risks breaking existing Jolie code though.

A preliminary implementation of `:=` is in https://github.com/jolie/jolie/tree/feature/deepcopy.

**Comment by thesave** [comment]
Just to clarify, following route 1 (deprecation), inline trees would preserve the current syntax, replacing `<<` with `:=`, and we would still be able (actually, we'd need to, in loops) to write things like `x[i] := y[j]` (so `x := y` desugars to `x[0] := y[0]`).
Anyhow, I like this route better than "silently" modifying the semantics of `<<` :)

**Reply by fmontesi** [reply]
Exactly! 🙂

**Reply by mwallnoefer** [reply]
Fine, nevertheless I would keep `<<` for appending data **outside** tree literals, since it resembles the common OS' shell syntax.

**Reply by fmontesi** [reply]
The problem is that it doesn't exactly append: whenever it finds the same node, it replaces it. Still, I agree it should be kept like this in case we go with the `:=` proposal.
